Review

buddazbs opened the pull request.

BuddazbsLab committed:
more update

BuddazbsLab committed:
last change

BuddazbsLab committed:
remove image

SergeyK2004 commented on src/components/article-params-form/BackgroundColorSelect.tsx:
```diff
@@ -0,0 +1,14 @@
+import { Select } from 'src/ui/select';
+import { backgroundColors } from 'src/constants/articleProps';
+
+export const BackgroundColorSelect = ({ value, onChange }: { value: string; onChange: (value: string) => void }) => {
```
Надо исправить:
Создание отдельных компонентов оберток над компонентами полей формы неоправданно. Каждый из созданных вами отдельных компонентов отличается от другого только тем, что в нем жестко прописано название. title является пропсом компонента Select и создавать отдельный компонент чтобы захардкордить в нем title не имеет смысла. 
Используйте компоненты Select  и RadioGroup ghb создании разметки в компоненте ArticleParamsForm передавая в них необходимые пропсы.

buddazbs replied:
Поправил

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,102 @@
+					<div className={styles.fieldBlock}>
+						<FontFamilySelect value={formState.fontFamily} onChange={(value) => handleChange('fontFamily', value)} />
+					</div>
+					<div className={styles.fieldBlock}>
```
Надо исправить:
Не нужно создавать множество дополнительных div, используйте компоненты полей формы, каждый из них и так самостоятельный блок.

buddazbs replied:
Поправил

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,102 @@
+	fontSize: string;
+};
+
+const defaultArticleStateValues: ArticleStateValues = {
```
**Надо исправить:**

Настройки статьи по умолчанию описаны в объекте `defaultArticleState`  в файле `constants/articleProps.ts` стартеркита. В проекте нужно использовать значения из этого объекта как настройки по умолчанию, как для самой статьи, так и как начальные значения полей форм.
Не допускается использовать для значения по умолчанию элементы с нулевым индексом из массивов свойств. То что настройки по умолчанию имеют именно нулевой индекс можно считать просто совпадением и это может измениться в любой момент.

buddazbs replied:
Поправил

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,102 @@
+	arrowButtonRef: React.RefObject<HTMLDivElement>;
+};
+
+type ArticleStateValues = {
```
Тип объекта с настройками прописан в src/constants/articleProps.ts 
Используйте уже имеющийся тип, а не создавайте еще один

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
 const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
+type ArticleStateValues = {
```
Надо исправить:
Тип объекта с настройками прописан в src/constants/articleProps.ts
Используйте уже имеющийся тип, а не создавайте несколько копий одного и того-же типа.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
+	fontSize: string;
+};
+
+const defaultArticleStateValues: ArticleStateValues = {
```
**Надо исправить:**

Настройки статьи по умолчанию описаны в объекте `defaultArticleState`  в файле `constants/articleProps.ts` стартеркита. В проекте нужно использовать значения из этого объекта как настройки по умолчанию, как для самой статьи, так и как начальные значения полей форм.
Не допускается использовать для значения по умолчанию элементы с нулевым индексом из массивов свойств. То что настройки по умолчанию имеют именно нулевой индекс можно считать просто совпадением и это может измениться в любой момент.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
+};
+
 const App = () => {
+	const [isSidebarOpen, setSidebarOpen] = useState(false);
```
**Надо исправить:**

Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния в родительский компонент не имеет смысла для нашего приложения. Состояние формы нужно описать в самом компонент `ArticleParamsForm`, а не передавать его из `App`, т.к. это состояние описывает именно `ArticleParamsForm`. В состоянии `App` должны хранится только данные о текущем отображении статьи.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
+	const arrowButtonRef = useRef<HTMLDivElement | null>(null);
+
+	const handleSidebarOpen = () => setSidebarOpen(true);
+	const handleSidebarClose = () => setSidebarOpen(false);
```
Надо исправить:
Не стоит создавать лишние сущности. Функция вызывающая функцию не имеет смысла. 
Обратите внимание, что даже 4 строками ниже, вы вместо вызова вот этой функции просто используете setSidebarOpen(false) в 52 строке.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
+		setSidebarOpen(false);
+	};
+
+	const handleReset = () => {
```
Надо исправить:
Все обработчики событий формы должны находиться в компоненте формы.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
+		setArticleState(defaultArticleStateValues);
+	};
+
+	useEffect(() => {
```
**Можно лучше:**

Будет здорово вынести закрытие по клику в оверлей в кастомный хук. Посмотрите, как пример, использование хука в компоненте Select. Только лучше будет использовать событие `mousedown`

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
-			<ArticleParamsForm />
+			}
+		>
+			<ArticleParamsForm
```
Надо исправить:
В компонент формы вам нужно передать только текущие настройки статьи и функцию изменения состояния.

buddazbs replied:
Поправил

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,102 @@
-				<form className={styles.form}>
+			<ArrowButton isOpen={isOpen} onClick={onToggle} ref={arrowButtonRef} />
+			<aside ref={asideRef} className={clsx(styles.container, { [styles.container_open]: isOpen })}>
+				<form className={styles.form} onSubmit={handleSubmit} onReset={handleReset}>
```
**Отлично:**

Отлично, что обработчик сохранения данных вызывается при отправке формы.

buddazbs replied:
Поправил

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,102 @@
-			<aside className={styles.container}>
-				<form className={styles.form}>
+			<ArrowButton isOpen={isOpen} onClick={onToggle} ref={arrowButtonRef} />
+			<aside ref={asideRef} className={clsx(styles.container, { [styles.container_open]: isOpen })}>
```
**Отлично:**

Здорово, что используете библиотеку `clsx`.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
+	fontSize: fontSizeOptions[0].value,
+};
+
 const App = () => {
```
**Можно лучше:**

Обычно корневой `index.tsx` держат максимально компактным. Стоит вынести компонент `App` в отдельный файл в папке `components/app/App.tsx`.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,31 +1,101 @@
+
 	return (
 		<main
 			className={clsx(styles.main)}
```
Можно лучше:

Для одиночных селекторов не стоит использовать библиотеку `clsx`

buddazbs replied:
Поправил

SergeyK2004 reviewed (changes requested):
Добрый день!  

У вас получилась отличная работа:
* Полностью реализован функционал.
* Сайдбар закрывается по клику вне сайдбара
* Обработчик события `submit` навешивается на форму, а не на клик по кнопке
* Для условного добавления стилей используется библиотека `clsx`
* Типизированы все пропсы


Но есть пара моментов, который надо поправить:

- За состояние страницы отвечает компонент `App`, а за состояние формы компонент `ArticleParamsForm`. Нужно продумать логику передачи данных от `App` в форму и обратно.
- Компонент страницы не должен знать про состояние формы. Состояние формы надо перенести в компонент `ArticleParamsForm`. Туда же нужно перенести все обработчики.
- Для дефолтных состояний статьи и формы используется `defaultArticleState`

Правки можно лучше остаются на ваше усмотрение:

- Обычно корневой `index.tsx` держат максимально компактным. Стоит вынести компонент `App` в отдельный файл в папке `components/app/App.tsx`.
- Для одиночных селекторов не стоит использовать библиотеку `clsx` (см. компоненты `Article` и `App`).
- Обработчик клика по оверлею для закрытия сайдбара лучше попробовать реализовать через кастомный хук.

Успешного рефакторинга!

BuddazbsLab committed:
feat(review): refactor sidebar architecture, improve form state and UX

Move App component to a separate file, keep index.tsx minimal
Store article state in App, form state and handlers in ArticleParamsForm
Use defaultArticleState and ArticleStateType for default values and typing
Remove unnecessary wrappers and field components, use Select and RadioGroup directly
Avoid clsx for single selectors
Implement custom hook useOverlayClose for sidebar overlay click handling
Fix typing and state transfer between components
UI and logic fully match project requirements and review checklist

buddazbs commented:
Внес правки по коду.

BuddazbsLab committed:
feat(fix): little fix

BuddazbsLab committed:
feat(fix): little fix

SergeyK2004 commented on src/components/app/App.tsx:
```diff
@@ -0,0 +1,30 @@
+
+export const App = () => {
+	const [articleState, setArticleState] = useState<ArticleStateType>(defaultArticleState);
+	const [isSidebarOpen, setSidebarOpen] = useState(false);
```
**Надо исправить:**

Состояние открытия формы необходимо создавать и хранить в самой форме. Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния формы в родительский компонент не имеет смысла для нашего приложения.

buddazbs replied:
Поправил

SergeyK2004 commented on src/components/app/App.tsx:
```diff
@@ -0,0 +1,30 @@
+	return (
+		<main className={styles.main}>
+			<Article articleState={articleState} />
+			<ArticleParamsForm
```
Надо исправить:
В компонент формы вам нужно передать всего два пропса. 
Текущее состояние статьи и функцию изменения состояния.
Все остальное это функционал самой формы и должно быть описано в компоненте формы.

buddazbs replied:
Поправил

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,102 @@
+	const handleReset = (e: React.FormEvent) => {
+		e.preventDefault();
+		setFormState(defaultArticleState);
+		onReset();
```
Надо исправить:
У статьи нет понятия событий (reset), просто установите состояние по умолчанию.

```ts
onApply(defaultArticleState);
```

buddazbs replied:
Поправил

SergeyK2004 commented on src/components/article/Article.tsx:
```diff
@@ -1,49 +1,86 @@
+}
+
+export const Article = ({ articleState }: ArticleProps) => {
+	const {
```
Надо исправить:
Компонент статьи не требует изменений.  Установка стилей производится в App, так-же как это было сделано в стартовом комплекте. Верните функционал компонента из стартового набора и изменяйте стили отображения в App, так-же как было в стартере, только используйте для этого ваш стэйт.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -13,20 +14,62 @@ const domNode = document.getElementById('root') as HTMLDivElement;
+				asideRef={sidebarRef}
+				arrowButtonRef={arrowButtonRef}
+			/>
+			<Article articleState={articleState} />
```
Надо исправить:
У вас стили установлены выше, в 54 строке. Передавать настройки отображения в компонент самой статьи не имеет смысла, так-же как и изменять компонент статьи.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -13,20 +14,62 @@ const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
 const App = () => {
```
Надо исправить:
У вас есть отдельно созданный компонент App и здесь вы продолжаете использовать написанный ранее. Определитесь какой компонент вы используете и оставьте только один экземпляр.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -13,20 +14,62 @@ const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
 const App = () => {
+	const [isSidebarOpen, setSidebarOpen] = useState(false);
```
Надо исправить:

Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния в родительский компонент не имеет смысла для нашего приложения. Состояние формы нужно описать в самом компонент ArticleParamsForm, а не передавать его из App, т.к. это состояние описывает именно ArticleParamsForm. В состоянии App должны хранится только данные о текущем отображении статьи.

buddazbs replied:
Поправил

SergeyK2004 commented on src/index.tsx:
```diff
@@ -13,20 +14,62 @@ const domNode = document.getElementById('root') as HTMLDivElement;
+	const handleSidebarClose = () => setSidebarOpen(false);
+	const handleSidebarToggle = () => setSidebarOpen((open) => !open);
+
+	const handleApply = (newState: ArticleStateType) => {
```
Надо исправить:
Компонент приложения не должен обрабатывать события формы. Собфтия формы должны существовать и обрабатываться только в компоненте формы.

buddazbs replied:
Поправил

SergeyK2004 reviewed (changes requested):
Очень хорошо, что вы исправили создание разметки формы используя компоненты UI. Но часть недочетов в App осталась без исправлений. Кроме того, у вас присутствует отдельной созданный компонент App, хотя используется написанный в index.ts в котором недочеты не исправлены.
Выполните исправление всех указанных недочетов и уберите все неиспользуемые компоненты из проекта.

BuddazbsLab committed:
feat(fix): Refactor: clean up project, remove unused components and legacy code

- Removed all unused components, hooks, and legacy files
- Centralized all article parameters in articleProps.ts
- Unified App component and fixed all previous issues
- Localized form and sidebar state in ArticleParamsForm
- Cleaned up all imports, types, and constants
- Applied best practices for React and TypeScript

buddazbs commented:
> Очень хорошо, что вы исправили создание разметки формы используя компоненты UI. Но часть недочетов в App осталась без исправлений. Кроме того, у вас присутствует отдельной созданный компонент App, хотя используется написанный в index.ts в котором недочеты не исправлены. Выполните исправление всех указанных недочетов и уберите все неиспользуемые компоненты из проекта.

Надеюсь правильно понял.

BuddazbsLab committed:
fix: repair change size Font

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,17 +1,105 @@
+			/>
+			<aside
+				className={clsx(styles.container, { [styles.container_open]: isOpen })}
+				style={{
```
Стили формы не изменяются динамически и устанавливаются в модуле. Применять здесь стили предназначенные для статьи некорректно

SergeyK2004 reviewed (approved):
Очень хорошо, все основные недочеты устранены.
Ваша работа принята.
Желаю успехов в дальнейшем освоении Реакта.
